### DISCUSSION TITLE
Custom Form Builder

### DISCUSSION THREAD

**Discussion by kasugaijin** [rubyforgood/pet-rescue#467] [Ideas]
**Background:** 
Each organization needs to know information about new adopters and fosterers who register. Often, the questions between organizations will differ, so an organization needs to be able to define its own questions.

The two user types that are asked questions after creating an account are adopters and fosterers - the questionnaire for each may overlap, but also be different. A user can also be both an adopter and fosterer. The responses to these form(s) make up their AdopterFosterProfile that is used by staff.

**Custom Form Builder:** 
- Each organization's staff can CRUD their own questions, defining the question and data-type. 
- Questions can be grouped into questionnaires (e.g., for new adopters, or new fosterers). 
- The adopter Questionnaire may overlap with the fosterer questionnaire. If a user already has responses for a question (e.g., an adopter also becomes a fosterer) these fields should be pre-populated so they do not need answering again.
- Responses to questions should be readily retrievable to populate the adopter/foster’s AdopterFosterProfile.
- Phase I - developers can CRUD questions and create a questionnaire (a grouping of questions)
- Phase II - integrate with the front-end so staff can CRUD questions/group into questionnaires, and new adopter/fosterers are shown an org’s questionnaire after account creation and the responses populate their AdopterFosterProfile

**Comment by ghost** [comment]
## Data Modeling

Here's one possible approach to modeling this.  I tried to take into account the context here and in the discussion in #388.

Edit: Changed `Form` -> `Questionnaire` and `type` -> `account_type` based on discussions.

![pet_rescue_form_modeling](https://github.com/rubyforgood/pet-rescue/assets/765993/d0d83a3b-0a25-4101-9bea-5e053e1eeafa)

### Questionnaire

The Questionnaire model encapsulates a group of questions around a specific topic (e.g. lifestyle, housing situation, etc).  It has a title and optional instructions defined by staff users that are displayed to adopter/fosterer users.

Questionnaires have a many-to-many relationship with Organizations using the `OrganizationQuestionnaire` model.  The join table specifies a few additional properties, including the type of join (e.g. `adopter`, `fosterer`), and the order in which the questionnaire should appear when viewing/editing a profile.  This relationship should enable this functionality:

> The adopter Questionnaire may overlap with the fosterer questionnaire.

### Question

The Question model is similar to what was originally proposed.  It has an `input_type` which'll be an enum in the app that maps similarly to Google Forms question types (short answer, long answer, choices, etc).  It has an `options` field that stores arbitrary settings that are specific to `input_type`, like choices, placeholder text, and so on.

Like Questionnaires, Questions also have a `sort_order` which drives the order in which questions are presented within a profile.

### Answer

The Answer model stores an answer to a question as provided by a User.  I messed around a bit with `Response` and `Profile` classes that encapsulated groups of `Answers` but this bit stuck out to me:

> If a user already has responses for a question (e.g., an adopter also becomes a fosterer) these fields should be pre-populated so they do not need answering again.

If we are having to map answers to Responses/AdopterAccounts/FostererAccounts I think there's more risk of ending up in a situation where we are missing FKs and leaving already answered questions out of certain profile types.

### Continuity

I read up on some of the discussion around how to deal with updating and removing questions.  Personally, I would prefer to use soft deletes and gently warning staff users about changing the meaning of questions, but I do also understand that users may not always use the app the way we intend.  Some well-placed frontend warnings can help mitigate people changing the meaning of questions, but there's also no harm in just having another field on the `Answer` table for dumping what the question was at the time it was answered by a user.

**Reply by kasugaijin** [reply]
Yes the question around continuity is a good point and not something I had considered. It could really cause some confusion if staff change the meaning of questions. I like the idea of soft deletion, and also recording the question when the answer was provided. That would allow one to make sense of the information. 

Another possible approach, after some thought, is to prevent editing of any questions and allow only creation and deletion. I think that is the safest approach to prevent mismatch of questions and answers.

Practically speaking, most orgs in existence should already have their questionnaires written down, so I think for these orgs not being able to edit questions should not pose too much of an issue. Their questions are alread well tested in the real world.

For any new orgs where this app is their first time, they may have a bit more trial and error...but I still think disabling editing is going to create the best outcomes in the long run with respect to user experience and data consistency.

**Reply by ghost** [reply]
Sounds good to me, good to start out safe and with limitations and then if there's demand add the functionality for it. 👍🏻

**Reply by maebeale** [reply]
THANKS for hopping on this @uoodsq !!!

- May we consider naming Form something like FormQuestion? 
At first glance I was thinking Form was performing OrganizationForm's function. Form is such a loaded word that I think the ergonomics for devs (esp open source devs) will go up if we avoid using it to mean a part of a Form.Additionally, my hope is that we move later to not first setting up an account, but having a (real-life) singular "Form" trigger all the work. But maybe I've missed something about what Form means to you!

- If possible, I recommend we avoid restricting edit after answers exist
I've got that back of the envelope drawing running in the wild and while orgs usually have their questions, they also have a lot of scenarios where they've got a typo or want to change format or word ordering etc. Esp when using a Form perennially, which most of these orgs will be doing. I'm fine with whatever is decided -- just wanted to chime in that it's far less often that the meaning changes than there is need for text changes that do not change the meaning.

**Reply by maebeale** [reply]
Why does the OrganizationForm need 'type' to identify uniqueness? (Wouldn't both FK's give us that already?)

**Reply by kasugaijin** [reply]
@maebeale I am fine with not restricting editing of a question. I think we need to make it clear on the UI that editing for typos and small changes is fine, but do not recommend any changes for question meaning...in that case make a new question.

**Reply by ghost** [reply]
> Why does the OrganizationForm need 'type' to identify uniqueness? (Wouldn't both FK's give us that already?)

Because a Form could be associated with an Organization more than once.

For example, if an org had a "Housing" form, and wanted to use it for both Adopter and Foster profiles, there would be two OrganizationForm entries with the same `form_id` and `organization_id`, but one would have `adopter` as its type and the other would have `foster`.  We could change `type` to `account_type` to make that more clear.

> May we consider naming Form something like FormQuestion?

I agree that Form isn't a great name here.  Would `FormSection`, `Topic`, or `Questionnaire` be more appropriate?  In the above diagram, `Form` is meant to encapsulate a group of questions around a topic (housing, lifestyle, etc), and OrganizationForm is meant to be the mapping of those groups onto Organizations.

> If possible, I recommend we avoid restricting edit after answers exist

I'm not worried about it either way, it doesn't impact the data model and it's just a few bits in the controller and view to support editing.  Some good alerts/tooltips can help guide the user in making good choices around editing questions.

**Reply by ghost** [reply]
Trying `Questionnaire` out.  LMK if that makes more sense. :)

**Reply by maebeale** [reply]
are we positive people will never have to fill out a form type twice, @kasugaijin ?

**Reply by maebeale** [reply]
@uoodsq i like FormSection best, bc then we can mix and match the sections on different Questionnaires (maybe that's what we call what the user actually sees)

**Reply by maebeale** [reply]
i think that's the gap for me -- the names should represent things wrt the admin views as well as what the applicant views, and those are different. for ex, if a Fosterer decides to become an Adopter, on their profile we'd only need to see the Housing section once, even tho it's displayed on both Adopter and Foster "Questionnaires" 

(but if the applicant is already logged in, we could prepopulate Housing answers on the 'Questionnaire' so they can change them there, and we can overwrite their current User data; assuming we want the answers to be ephemera and the User model to store the latest-and-greatest contact info. i'm teetering YAGNI here, but, this is how it works in the setup i've had in the wild for 10yrs)

**Reply by kasugaijin** [reply]
@maebeale  
So an adopter or a fosterer has to sign up as just one of these user types. We won't have the ability to create a user with both types for MVP. So, if an adopter also then also becomes a fosterer, or a fosterer also then becomes an adopter, they would most likely be exposed to the same questionnaire again e.g., contact details, or using Bryan's example for housing, above. 

With regards to filling out a form type twice, I guess this depends on how users actually use the app. And in cases, won't be how we expect :)

My assumption is that it would make sense to group questions into logical, related units of questionnaires, so a group of questions for contact details, and another group for housing, for example. Now, assuming these questionnaires are usable for both adopter and fosterers, then I would say, yes, we will have cases where the user will be exposed to the same questionnaire more than once. This makes it easy for us to re-use answers so the user does not have to fill the same questionnaire out twice.

However, I think there might be times where an org might be inclined to create two separate Housing questionnaires, both with overlap, but also some differences, for adopters and fosterers. In this case, both questionnaires will need to be answered. It would be great to have the overlapping/identical questions that already have answers, answered. But I think this starts to get a lot more complex in terms of data management, and probably not a path we want to go down. So, perhaps we might consider prescribing some rules around developing questionnaires for orgs? I.e., develop questionnaires that can be used in both cases. And then, any questions that are unique to adopters or fosterers should live a separate questionnaire(s). Does this make sense? I feel like I am rambling a bit.


In terms of UX, I was just thinking about how we might display this and get the answers that we need. In the app currently, a user signs up as an adopter and then redirected to fill out the adopter profile form. Once they do this, they have access to their adopter profile, which is comprised of the form answers. 

With this new implementation, I was thinking we would no longer have the form and profile being separate. Instead, the user can always access their profile, and depending on the user type, we display the questionnaires they need to answer, so it's basically a page of independent forms. Once they have them all answered their profile is considered 'complete' and they can start to apply on pets etc. Apologies if I am just playing catch up here :)

**Reply by ghost** [reply]
If I'm reading this thread correctly, it sounds like the data model being proposed here should work.  I'm not worried about settling on naming conventions here, and I'm going to start putting together some wireframes for how the UX of this should work which should give us something more concrete to focus on as to how admins and users will interact with this. :)

**Reply by kasugaijin** [reply]
Yes I think so, @uoodsq !

**Reply by elight** [reply]
New guy here with some possibly naive questions and a few thoughts, after looking some at this schema and schema.rb.

About this schema: how would this represent a list of people who filled out a given Questionnaire for an Org?  It looks like it'd have to be a multi-way join of User -> Answer -> Question -> Questionnaire -> OrgQuestionnaire?

Should OrgQuestionnaire belong to AdopterApplication?

At the risk of over-stepping, I wonder about the ~~unity~~ utility of re-using individual questions, within an organization. What is the use case for this? I understand that there will be some re-use between adopter and fosterer questionnaires. Would it be easier, from a UX and modeling perspective, to offer an abstraction that bundles sets of questions together such that one could drop in a bank of reused questions, such as "Previous Pet Ownership Experiences" which could contain multiple questions?

**Reply by kasugaijin** [reply]
A PR was merged for the building blocks: https://github.com/rubyforgood/pet-rescue/pull/629 with some renaming
I think this approach gives more flexibility for an org to define and group questions together?

**Reply by ghost** [reply]
@elight Great questions, thanks.  As @kasugaijin mentioned, names of things ended up changing in what got merged.

> how would this represent a list of people who filled out a given Questionnaire for an Org?

I suspect the query to generate this list of people would be pretty gnarly, or it'd end up being something that got assembled Ruby-side.  This particular use case hasn't come up yet, nor was performance a target, so I opted to try to keep the schema on the simpler side (whether or not that ended up happening is debatable 😆)

The use case I had more in mind was for rendering individual profiles, so given a specific user and profile type, iterate over the `Forms` and then the `Questions`, rendering either the `Question` as a form input, or loading the user's `Answer` and displaying it.

> Would it be easier, from a UX and modeling perspective, to offer an abstraction that bundles sets of questions together such that one could drop in a bank of reused questions, such as "Previous Pet Ownership Experiences" which could contain multiple questions?

Yes!  I think #629 addresses this.  `Form` encapsulates a group of questions like you describe.  `FormProfile` "attaches" the `Form` to a given profile type ("adopter" or "fosterer") with a sort order.

**Comment by ghost** [comment]
## Question

What should happen to existing data in the `AdopterProfile` model?

I don't think it would be out of the question to attempt to model the static Q&As in that table into this new model, and migrate the data over.

**Reply by kasugaijin** [reply]
So the app is purely in development. I think we can get rid of any redundant tables without any need for handling existing data.

**Reply by ghost** [reply]
Sounds good.  We could still seed new Organizations with some boilerplate questions.

**Comment by ghost** [comment]
## Question

> Responses to questions should be readily retrievable to populate the adopter/foster’s AdopterFosterProfile.

Is there some process or something I don't understand where there needs to be a "population" of a profile?  Or does this new data model supercede the existing `AdopterProfile` (and presumably a `FostererProfile`)?  In the proposed model, a "profile" would become the collection of answers for the given set of forms that make up the "adopter" or "fosterer" form, and wouldn't need to be copied into some other db table.

**Reply by kasugaijin** [reply]
Yes you are right. We don't necessarily need a new table. Perhaps there may be value in some class that collates this data for a given user? But it can pull from the Answers table without any issues I think. So, in that sense, the AdopterFosterProfiles table would probably become redundant.

**Reply by ghost** [reply]
To me, adding that type of table feels like something that should be done if the app can't express what it needs out of the tables with all of what ActiveRecord has to offer in a performant manner.

**Reply by maebeale** [reply]
Agree! Esp since the same humans can be both. Going to a User Profile, seeing the answers they have can be grouped by FormSection/Form/QuestionFrom/FormQuestion in the display. And if really we only want to see a particular one at a time, we can use a url param from whatever they came to hide/show the parts we want visible for that purpose.

**Reply by ghost** [reply]
> FormSection/Form/QuestionFrom/FormQuestion

🤣 names are hard

**Comment by nsiwnf** [comment]
Was looking at the [Baja Pet Rescue adopter form](https://github.com/kasugaijin/baja-pet-rescue/blob/main/app/views/adopter_profiles/_form.html.erb) - is it worth building in some "conditional" logic into questions? 
Ex. 
"Will dog ownership be regularly shared?" Yes/No
If Yes -> "Briefly describe who else will share dog ownership"

**Reply by kasugaijin** [reply]
Yea this is a good question. I think it would add value to have the ability for a question to show or hide depending on the answer to its parent question that it belongs to. 

Thinking about it quickly, to facilitate this we would have to enable staff to create a question that belongs to another ie the parent and child. In practical terms, the parent would be a Boolean, or maybe a select, and we could use Hotwire to render the child question depending on the selection made on the parent question. Does this pass the sniff test?

**Reply by ghost** [reply]
I worry that modeling dependencies between questions could get messy.  What if you encapsulated the boolean and details fields in a single Question with a custom input type? (e.g. "Yes/No with details")

You could still ship whatever frontend bits are needed to switch the fields on/off, but the answers would get stored in a single Answer row.

**Reply by kasugaijin** [reply]
Ah yes this makes far more sense, I prefer that approach @uoodsq

**Comment by ghost** [comment]
I've put together some wireframes to illustrate the experience for staff managing forms and questions:

A new section on the dashboard sidebar ("Forms") takes you to a page where you can see all Forms that exist for a given organization.  The table indicates how a Form is used (Adopter profiles, Fosterer profiles, or both):

![Show Forms](https://github.com/rubyforgood/pet-rescue/assets/765993/83841b3f-f9fb-475d-b919-f76f531b1311)

Adding/editing a Form, including selecting the profiles to which the Form applies:

![New Form](https://github.com/rubyforgood/pet-rescue/assets/765993/60232bd0-cf80-49fb-ae65-3bb66cebfef4)

Once a Form is created, it can be visited to see the Questions that exist in it:

![Show Questions](https://github.com/rubyforgood/pet-rescue/assets/765993/7c380b9c-e9f6-410f-9792-722466a321b4)

Adding a new Question:

![New Question (Short Answer)](https://github.com/rubyforgood/pet-rescue/assets/765993/9e26758b-a4da-40e3-a1be-f59cb4d896b9)

Changing a Question's type drives the options that are presented to the staff user.  For example, changing to multiple choice allows the user to enter the choices presented to the end user:

![New Question (Multiple Choice)](https://github.com/rubyforgood/pet-rescue/assets/765993/dd88ce35-436a-4892-b0c1-c48067981e40)

Once Questions are added, a separate page exists that allows the staff user to drag/drop and reorder the Questions within the Form:

![Reorder Questions](https://github.com/rubyforgood/pet-rescue/assets/765993/a5d74b56-6f7e-4b98-bf8f-2dda60ce9f78)

Similarly, Forms can be reordered within a profile type (adopter, fosterer):

![Reorder Forms](https://github.com/rubyforgood/pet-rescue/assets/765993/dfa459e1-4896-4352-875c-083d3729dc44)

**Reply by kasugaijin** [reply]
This is awesome, thank you @uoodsq 
We had our project weekly tonight and went over what you have provided, and it all makes sense. I for one am happy to see how it might look, and it all looks very straightforward and logical to me.

We had a couple of discussion points come up:

- Editing/deleting questions by org staff - in the discussion above we agreed that org staff should be able to edit questions, but provide a warning that meaning should not be changed. What should we do if org staff delete a question? I agree with your proposal to soft delete answers. Do we also want to save the question text on the response to provide contextm in case a question is edited (and inadvertantly meaning changes)?
- If an adopter or fosterer user completes their profile questionnaires, and then staff add another question to a form, or form of questions at a later date, the adopter or fosterer will need to be notified they have questions to complete.
- How does drag and drop work on smart device/touch screen? I have not worked in drag an drop in rails, nor for smart devices.
- Is there a necessity for a ‘required’ form? My thinking is no. A form is something all users will need to answer (save for the user specific adopter/foster form which only each user type answers).
- number of forms - so each response to a form question will be on the same model right? So is it feasible to have a separate form/submit for each form? Ie a user can just fill out the housing form and submit it? On a related point, how can we handle validations? I am not quite understanding if we can have back end model validations or not. I am used to using them when we know the form fields and attributes being used. But in this case we don’t know them ahead of time. This also relates to the separate form submissions, if we submit only part of a models attributes will validations for other attributes fail? I am sure there’s ways to solve for this but it’s new to me :)
- to @nsiwnf point above, some questions will have associated questions that e.g., anserwing Yes to 'do you have pets at home' would then show its related question of 'describe what and how many'.

**Reply by ghost** [reply]
> Editing/deleting questions by org staff - in the discussion above we agreed that org staff should be able to edit questions, but provide a warning that meaning should not be changed. What should we do if org staff delete a question? I agree with your proposal to soft delete answers. Do we also want to save the question text on the response to provide contextm in case a question is edited (and inadvertantly meaning changes)?

Yep, I think soft deletes + a `question_snapshot` on the `Answer` model will be a good way of capturing what the state of the question was at the time the answer was made without breaking the logical relationship between the question and answer.

> If an adopter or fosterer user completes their profile questionnaires, and then staff add another question to a form, or form of questions at a later date, the adopter or fosterer will need to be notified they have questions to complete.

Agreed.  Not sure what the mechanism for this looks like yet.  Could be a callback whenever a new question or form is created, or a job that goes and looks for incomplete profiles.

> How does drag and drop work on smart device/touch screen? I have not worked in drag an drop in rails, nor for smart devices.

Desktop browsers can use [drag and drop events](https://developer.mozilla.org/en-US/docs/Web/API/HTML_Drag_and_Drop_API) which I believe are supported even in mobile browsers, although I'll have to research more into the mobile side to see how they work specifically.  There are also [touch events](https://developer.mozilla.org/en-US/docs/Web/API/Touch_events) available in mobile browsers too.  There do exist Rails plugins that can manage sort order but to be honest it's just an integer column + a controller action that accepts a new sort order from the frontend.

> Is there a necessity for a ‘required’ form? My thinking is no. A form is something all users will need to answer (save for the user specific adopter/foster form which only each user type answers).

Agreed.  I suppose if all questions in a form were optional then it would follow that the form is optional, but I don't think an explicit `required` field is needed on forms.

> number of forms - so each response to a form question will be on the same model right? So is it feasible to have a separate form/submit for each form?

It can be done either way, the controller used to accept user answers can be structured to accept one form at a time or all applicable forms at once.  There's just a single `answers` table that stores answers for all questions regardless of what form they are associated with.

> On a related point, how can we handle validations? I am not quite understanding if we can have back end model validations or not.

We can hook whatever custom logic we want up to `ActiveRecord` validations.  The `Answer` model will need a validator that looks at the `input_type`, `required`, and `options` of its associated `Question` and decide whether the `value` its about to store is valid or not.

> This also relates to the separate form submissions, if we submit only part of a models attributes will validations for other attributes fail?

Creating a batch of `Answer` models will be done in a transaction, so if one `Answer` fails to validate regardless of whether or not its in a single form or group of forms, the whole batch will roll back and the user will be presented error messages to correct.

> to @nsiwnf point above, some questions will have associated questions that e.g., anserwing Yes to 'do you have pets at home' would then show its related question of 'describe what and how many'.

Yes, if those sorts of questions can be generalized to an input type (e.g. "Yes/No + Explanation")  then this data model should still work fine, a question's `options` could be extended to support custom labels and such.

**Reply by kasugaijin** [reply]
@uoodsq thank you for your responses - all makes sense! Please let me know how I/we can facilitate next steps on this work.

**Comment by mononoken** [comment]
> Once Questions are added, a separate page exists that allows the staff user to drag/drop and reorder the Questions within the Form:

I was thinking, maybe we should consider adding buttons as well as an alternative way to adjust sorting. One possible version would be an up and down button on each question to move the questions. I like the intuitiveness of click-and-drag, but I think another more accessible option would be an improvement.

**Reply by kasugaijin** [reply]
Yes I think we can discuss in more detail how we think this should be implemented for MVP. My gut is we go with one solution for MVP, to keep things streamlined and moving, and that solution ideally works on desktop and mobile alike. In the case of drag and drop, I think we should get user feedback before building more than we need to. If what we build for drag and drop for MVP works well, win, if not, we can digest feedback and solve any problems.
